### PR TITLE
Jormun: Deactivate min_car and min_bike for journeys without walking mode

### LINKED ISSUE #2235: Desactivate min_mode rules when we have only one fallback mode
If we ask for a journey with `first_section_mode[]=bike` and that the solution have less than `min_bike` bike duration, we will filter every solution. Same with car.

These constraints of minimal duration only have sense if we have another fallback mode (as walking). They must be deactivated in such cases.

Googlegroup conversation: https://groups.google.com/forum/#!topic/navitia/JxR2Eo-lHeY

### PR DESCRIPTION
**Pull request by azime**
Fix #2235

### CODE REVIEW

**Comment by buildbotctp** [comment]
Build finished.

File: source/jormungandr/jormungandr/scenarios/utils.py [hunk 1]
```diff
@@ -254,15 +254,20 @@ def updated_common_journey_request_with_default(request, instance):
...
 
+def walking_in_origin_destination_mode(request):
+    return 'walking' in request['origin_mode'] and 'walking' in request['destination_mode']
+
+
 def updated_request_with_default(request, instance):
     updated_common_journey_request_with_default(request, instance)
 
-    if request['_min_car'] is None:
+    if request['_min_car'] is None and walking_in_origin_destination_mode(request):
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/utils.py]: So we apply the min only if there is walking in both origin AND destination?
@TeXitoi I think your opinion is welcome on that.

File: source/jormungandr/tests/journey_common_tests.py [hunk 2]
```diff
@@ -1208,3 +1207,50 @@ def test_strange_line_name(self):
...
+        query = '{}&last_section_mode[]=car&first_section_mode[]=car&' \
+                'last_section_mode[]=walking&first_section_mode[]=walking'.format(sub_query)
+        response = self.query_region(query)
+        self.is_valid_journey_response(response, query)
+        assert len(response['journeys']) == 1
+        assert len(response['journeys'][0]['sections']) == 1
+        assert response['journeys'][0]['sections'][0]['mode'] == 'walking'
+        assert response['journeys'][0]['sections'][0]['duration'] == 37
+
+        query = '{}&last_section_mode[]=bike&last_section_mode[]=walking&' \
```
Comment (pbougue) [inline source/jormungandr/tests/journey_common_tests.py]: Depending on the opinion for the other remark, you should probably add a test with only walking activated at beginning side.

File: source/jormungandr/tests/journey_common_tests.py [hunk 3]
```diff
@@ -1208,3 +1207,50 @@ def test_strange_line_name(self):
         self.is_valid_journey_response(response, query)
 
         assert len(response['journeys']) == 1
+
+@dataset({"main_routing_test": {}})
+class JourneyMiniBikeMiniCar(object):
```
Comment (TeXitoi) [inline source/jormungandr/tests/journey_common_tests.py]: "mini bike" is quite disturbing... 😉 (petit petit petit, tout est mini dans notre vie !)

File: source/jormungandr/tests/journey_common_tests.py [hunk 4]
```diff
@@ -1208,3 +1207,50 @@ def test_strange_line_name(self):
...
+        """
+        journeys from B to D
+        """
+        b_coord = "0.0002694935945864127;8.98311981954709e-05"
+        d_coord = "0.0002694935945864127;0.0"
+        sub_query = 'journeys?from={from_coord}&to={to_coord}&datetime={datetime}'.format(from_coord=b_coord,
+                                                                                          to_coord=d_coord,
+                                                                                          datetime="20120614T080000")
+        # Deactivate mini_bike and mini_car
+        query = '{}&last_section_mode[]=bike&first_section_mode[]=bike'.format(sub_query)
```
Comment (TeXitoi) [inline source/jormungandr/tests/journey_common_tests.py]: must also be tested with
  * `'{}&last_section_mode[]=bike`
  * `'{}&first_section_mode[]=bike'`
  * `'{}&last_section_mode[]=walking&first_section_mode[]=bike'`
  * `'{}&last_section_mode[]=bike&first_section_mode[]=walking'`

File: source/jormungandr/tests/journey_common_tests.py [hunk 5]
```diff
@@ -1208,3 +1207,50 @@ def test_strange_line_name(self):
...
+        response = self.query_region(query)
+        self.is_valid_journey_response(response, query)
+        assert len(response['journeys']) == 1
+        assert len(response['journeys'][0]['sections']) == 3
+        assert response['journeys'][0]['sections'][0]['mode'] == 'car'
+        assert response['journeys'][0]['sections'][0]['duration'] == 2
+
+        # Activate mini_bike and min_car
+        query = '{}&last_section_mode[]=car&first_section_mode[]=car&' \
+                'last_section_mode[]=walking&first_section_mode[]=walking'.format(sub_query)
```
Comment (TeXitoi) [inline source/jormungandr/tests/journey_common_tests.py]: must also be tested with
  * `'{}&last_section_mode[]=car&last_section_mode[]=walking`
  * `'{}&first_section_mode[]=car&first_section_mode[]=walking`

File: source/jormungandr/tests/journey_common_tests.py [hunk 6]
```diff
@@ -1208,3 +1207,50 @@ def test_strange_line_name(self):
...
+        sub_query = 'journeys?from={from_coord}&to={to_coord}&datetime={datetime}'.format(from_coord=b_coord,
+                                                                                          to_coord=d_coord,
+                                                                                          datetime="20120614T080000")
+        # Deactivate mini_bike and mini_car
+        query = '{}&last_section_mode[]=bike&first_section_mode[]=bike'.format(sub_query)
+        response = self.query_region(query)
+        self.is_valid_journey_response(response, query)
+        assert len(response['journeys']) == 1
+        assert len(response['journeys'][0]['sections']) == 1
+        assert response['journeys'][0]['sections'][0]['mode'] == 'bike'
```
Comment (TeXitoi) [inline source/jormungandr/tests/journey_common_tests.py:1221]: the last section must also be checked when we test the first one.

**Review by TeXitoi — DISMISSED** [review]

File: source/jormungandr/jormungandr/scenarios/utils.py (hunk 1 above)
Comment (TeXitoi) [reply source/jormungandr/jormungandr/scenarios/utils.py]: yes, including the default values

**Comment by buildbotctp** [comment]
Build finished.

**Comment by buildbotctp** [comment]
Build finished.

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 7]
```diff
@@ -192,6 +192,14 @@ def _filter_similar_journeys(journeys, request, similar_journey_generator):
                           .format(other=j1.internal_id if worst == j2 else j2.internal_id))
 
 
+def activate_mini_duration(current_section, journey, request,  mini_duration):
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: I'd name the method `check_min_duration` (and make it return True if there's no filter to apply)

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 8]
```diff
@@ -192,6 +192,14 @@ def _filter_similar_journeys(journeys, request, similar_journey_generator):
...
+    if current_section == journey.sections[0]:
+        return 'origin_mode' in request and 'walking' in request['origin_mode'] \
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: Does that include the default value to walking?

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 9]
```diff
@@ -210,15 +218,19 @@ def _filter_too_short_heavy_journeys(journeys, request):
                 on_bss = False
             if s.type != response_pb2.STREET_NETWORK:
                 continue
-            if s.street_network.mode == response_pb2.Car and s.duration < request['_min_car']:
+
+            if s.street_network.mode == response_pb2.Car and activate_mini_duration(s, journey,
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: minor: indent on each "and"

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 10]
```diff
@@ -210,15 +218,19 @@ def _filter_too_short_heavy_journeys(journeys, request):
...
+                                                                                    request, request['_min_car']):
                 logger.debug("the journey {} has not enough car, we delete it".format(journey.internal_id))
                 mark_as_dead(journey, "not_enough_car")
                 break
-            if not on_bss and s.street_network.mode == response_pb2.Bike and s.duration < request['_min_bike']:
+            if not on_bss and s.street_network.mode == response_pb2.Bike \
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: minor: indent on each "and"

File: source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py [hunk 11]
```diff
@@ -948,3 +952,166 @@ def test_heavy_journey_bss():
 
     assert 'to_delete' not in journey.tags
 
+
+def test_activate_deactivate_mini_bike():
+    """
+
+      A                 B                           C            D
+      *................*---------------------------*.............*
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py]: minor: you can change `-----` to `======` for PT it's a bit more visual :) also the case elsewhere in this PR

File: source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py [hunk 12]
```diff
@@ -948,3 +952,166 @@ def test_heavy_journey_bss():
 
     assert 'to_delete' not in journey.tags
 
+
+def test_activate_deactivate_mini_bike():
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py]: sed s/mini/min/g everywhere in this PR ;)

File: source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py [hunk 13]
```diff
@@ -948,3 +952,166 @@ def test_heavy_journey_bss():
...
+    journey.sections[-1].street_network.mode = response_pb2.PUBLIC_TRANSPORT
+    journey.sections[-1].duration = 35
+
+    journey.sections.add()
+    journey.sections[-1].type = response_pb2.STREET_NETWORK
+    journey.sections[-1].street_network.mode = response_pb2.Bike
+    journey.sections[-1].duration = 7
+
+    journey_filter._filter_too_short_heavy_journeys([journey], request)
+    assert 'to_delete' not in journey.tags
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py:988]: Not sure this is how we want to behave by default: depends on if we populate with walking when nothing is provided. If we populate, we probably don't want to fix that in test (but that should be OK).
If we don't populate, we don't want that behavior (want to consider it's walking).
> ☑️ populated (tested)

File: source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py [hunk 14]
```diff
@@ -948,3 +952,166 @@ def test_heavy_journey_bss():
...
+
+    # case 2: request without origin_mode
+    journey.sections[-1].duration = 15
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py:991]: minor: Would be easier to read and understand if you only change `_min_bike` value (same for all tests almost - maybe create 2 journeys)
> Not easy to do

File: source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py [hunk 15]
```diff
@@ -948,3 +952,166 @@ def test_heavy_journey_bss():
...
+    journey_filter._filter_too_short_heavy_journeys([journey], request)
+    assert 'to_delete' not in journey.tags
+
+    # case 4: request without walking in origin_mode
+    journey.sections[0].duration = 5
+    journey.sections[-1].duration = 15
+    request = {'_min_bike': 8, 'origin_mode': ['bike']}
+
+    journey_filter._filter_too_short_heavy_journeys([journey], request)
+    assert 'to_delete' not in journey.tags
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py:1011]: We should also (it's a more realistic case if params are populated with default values) test with `{'_min_bike': 8, 'origin_mode': ['bike'], 'destinatin_mode': ['bike', 'walking']}` and that should be deleted if both are under min_bike

File: source/jormungandr/tests/journey_common_tests.py [hunk 16]
```diff
@@ -1208,3 +1207,108 @@ def test_strange_line_name(self):
...
+    def test_first_section_mode_bike_and_last_section_mode_walking(self):
+        query = '{sub_query}&last_section_mode[]=bike&first_section_mode[]=walking&' \
+                'datetime={datetime}'.format(sub_query=sub_query, datetime="20120614T080000")
+        response = self.query_region(query)
+        self.is_valid_journey_response(response, query)
+        assert len(response['journeys']) == 2
+        assert len(response['journeys'][0]['sections']) == 3
+        assert response['journeys'][0]['sections'][0]['mode'] == 'walking'
+        assert response['journeys'][0]['sections'][0]['duration'] == 17
+        assert response['journeys'][0]['sections'][2]['mode'] == 'bike'
```
Comment (pbougue) [inline source/jormungandr/tests/journey_common_tests.py]: minor: maybe `['sections'][-1]`

File: source/jormungandr/tests/journey_common_tests.py [hunk 17]
```diff
@@ -1208,3 +1207,108 @@ def test_strange_line_name(self):
...
+        assert response['journeys'][0]['sections'][2]['duration'] == 21
+
+        assert len(response['journeys'][1]['sections']) == 1
+        assert response['journeys'][1]['sections'][0]['mode'] == 'walking'
+        assert response['journeys'][1]['sections'][0]['duration'] == 276
+
+    def test_last_section_mode_bake_only(self):
```
Comment (pbougue) [inline source/jormungandr/tests/journey_common_tests.py]: typo: b`i`ke_only

File: source/jormungandr/tests/journey_common_tests.py [hunk 18]
```diff
@@ -1208,3 +1207,108 @@ def test_strange_line_name(self):
...
+                'datetime={datetime}'.format(sub_query=sub_query, datetime="20120614T080000")
+        response = self.query_region(query)
+        self.is_valid_journey_response(response, query)
+        assert len(response['journeys']) == 1
+        assert len(response['journeys'][0]['sections']) == 1
+        assert response['journeys'][0]['sections'][0]['mode'] == 'bike'
+        assert response['journeys'][0]['sections'][0]['duration'] == 62
+
+    def test_first_section_mode_bike_only(self):
+        query = '{sub_query}&first_section_mode[]=bike&datetime={datetime}'.format(sub_query=sub_query,
```
Comment (pbougue) [inline source/jormungandr/tests/journey_common_tests.py:1267]: you should also test with only `first=[bike, walking]` and there should be a filtering (the idea is to test more the default value, I may miss some cases also)

File: source/jormungandr/jormungandr/scenarios/journey_filter.py (hunk 7 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/journey_filter.py]: ok

File: source/jormungandr/jormungandr/scenarios/journey_filter.py (hunk 9 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/journey_filter.py]: ok

File: source/jormungandr/jormungandr/scenarios/journey_filter.py (hunk 10 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/journey_filter.py]: ok

File: source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py (hunk 12 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py]: :+1:

File: source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py (hunk 11 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py]: ok

File: source/jormungandr/tests/journey_common_tests.py (hunk 16 above)
Comment (azime) [reply source/jormungandr/tests/journey_common_tests.py]: ok

File: source/jormungandr/tests/journey_common_tests.py (hunk 17 above)
Comment (azime) [reply source/jormungandr/tests/journey_common_tests.py]: ok

File: source/jormungandr/tests/journey_common_tests.py (hunk 18 above)
Comment (azime) [reply source/jormungandr/tests/journey_common_tests.py:1267]: ok

Comment (azime) [reply source/jormungandr/tests/journey_common_tests.py]: ok

File: source/jormungandr/jormungandr/scenarios/journey_filter.py (hunk 10 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/journey_filter.py]: ok

File: source/jormungandr/jormungandr/scenarios/journey_filter.py (hunk 9 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/journey_filter.py]: ok

File: source/jormungandr/jormungandr/scenarios/journey_filter.py (hunk 8 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/journey_filter.py]: Yes

File: source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py (hunk 15 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py:1011]: [No](https://github.com/CanalTP/navitia/pull/2242/files#diff-0832703183fb3233e60448ac96ec726bR1290) ?

File: source/jormungandr/tests/journey_common_tests.py (hunk 6 above)
Comment (azime) [reply source/jormungandr/tests/journey_common_tests.py:1221]: In unit tests

File: source/jormungandr/tests/journey_common_tests.py (hunk 4 above)
Comment (azime) [reply source/jormungandr/tests/journey_common_tests.py]: ok

**Comment by buildbotctp** [comment]
Build finished.

**Comment by buildbotctp** [comment]
Build finished.

File: source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py (hunk 15 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py:1011]: ok

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 19]
```diff
@@ -192,6 +192,14 @@ def _filter_similar_journeys(journeys, request, similar_journey_generator):
                           .format(other=j1.internal_id if worst == j2 else j2.internal_id))
 
 
+def exceed_min_duration(current_section, journey, request,  mini_duration):
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: still some "mini" in the PR (ctrl + F)

File: source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py [hunk 20]
```diff
@@ -948,3 +952,197 @@ def test_heavy_journey_bss():
...
+    journey.sections[-1].duration = 5
+    request = {'_min_bike': 8, 'destination_mode': ['bike']}
+
+    journey_filter._filter_too_short_heavy_journeys([journey], request)
+    assert 'to_delete' not in journey.tags
+
+    # case 6: request with bike only in origin_mode destination_mode
+    journey.sections[0].duration = 15
+    journey.sections[-1].duration = 14
+    request = {'_min_bike': 8, 'origin_mode': ['bike'], 'destination_mode': ['bike']}
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py]: You have to increase that min for the test to be useful

File: source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py [hunk 21]
```diff
@@ -948,3 +952,197 @@ def test_heavy_journey_bss():
...
+    journey.sections[-1].duration = 5
+    request = {'_min_car': 8, 'destination_mode': ['car']}
+
+    journey_filter._filter_too_short_heavy_journeys([journey], request)
+    assert 'to_delete' not in journey.tags
+
+    # case 6: request with car only in origin_mode destination_mode
+    journey.sections[0].duration = 15
+    journey.sections[-1].duration = 14
+    request = {'_min_car': 8, 'origin_mode': ['car'], 'destination_mode': ['car']}
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py]: You have to increase that min for the test to be useful

File: source/jormungandr/tests/journey_common_tests.py (hunk 18 above)
Comment (pbougue) [reply source/jormungandr/tests/journey_common_tests.py:1267]: ☑️ done

**Comment by buildbotctp** [comment]
Build finished.

File: source/jormungandr/jormungandr/scenarios/journey_filter.py (hunk 19 above)
Comment (azime) [reply source/jormungandr/jormungandr/scenarios/journey_filter.py]: ok

**Comment by buildbotctp** [comment]
Build finished.

**Comment by buildbotctp** [comment]
Build finished.

**Review by pbougue — APPROVED** [review]

**Pull request merged**